Power Rework

collinlunn opened the pull request:
Working towards HackMD PJB power proposal, this is a complete rewrite of powernet code. Uses art assets from #715.

Feature List:
- [x] Different wires that do not connect with each other
- [x] Support for different strategies to find wires to connect to
- [x] Power Suppliers/Consumers
- [x] Smes/Substation
- [x] APC, auto-wire
- [x] Cleanup old power components in code and map

unusualcrow commented:
thanks

collinlunn commented:
Code for stuff other than remote powering with APCs + LV wire is functional and review-able.

ComicIronic commented on Content.Server/GameObjects/Components/Power/NewPower/BasePowerNetConnector.cs:
```diff
@@ -0,0 +1,151 @@
+            //otherwise make own powernet
+            if (PowerNet == null)
+            {
+                Debug.Assert(TrySetPowerNet(new PowerNet(this)));
```
The docs aren't super clear, but I believe this whole statement, including the arguments, will be elided in release builds. Is that deliberate?

ComicIronic commented on Content.Server/GameObjects/Components/Power/NewPower/BasePowerNetConnector.cs:
```diff
@@ -0,0 +1,151 @@
+        /// <returns>Returns whether or not this could join the supplied powernet.</returns>
+        public bool TrySetPowerNet(PowerNet powerNet)
+        {
+            if (TryJoinPowerNet(powerNet))
```
This is a confusing naming, at least for me. I would expect `TryJoin` to cause the component to join the net. Calling `Leave` *after* that, but before actually joining, seems very unintuitive. Would it make sense to change `TryJoin` to `CanJoin`?

ComicIronic commented on Content.Server/GameObjects/Components/Power/NewPower/BatteryComponent.cs:
```diff
@@ -0,0 +1,91 @@
+            {
+                CurrentCharge = MaxCharge;
+            }
+            UpdateStorageState();
```
You could replace the above 4 lines with `CurrentCharge = CurrentCharge`, I think. That's a question of personal style, though.

ComicIronic commented on Content.Server/GameObjects/Components/Power/NewPower/BatteryDischargerComponent.cs:
```diff
@@ -0,0 +1,90 @@
+        ///     power to its powernet.
+        /// </summary>
+        [ViewVariables]
+        private bool _hasCharge;
```
Is there a reason for this to be its own var, instead of just checking `_battery.BatteryState != BatteryState.Empty`?

ComicIronic commented on Content.Server/GameObjects/Components/Power/NewPower/BatteryStorageComponent.cs:
```diff
@@ -0,0 +1,87 @@
+        ///     Whether or not the battery is full, this will draw power
+        ///     if the battery is not full.
+        /// </summary>
+        private bool _needsPower;
```
Same comment about this needing to be a field.

ComicIronic commented on Content.Server/GameObjects/Components/NodeNetwork/INodeComponent.cs:
```diff
@@ -0,0 +1,125 @@
+        private IEnumerable<NodeComponent> GetReachableCompatibleNodes()
+        {
+            return GetReachableNodes().Where(node => node.NodeGroupID == NodeGroupID)
+                .Where(node => node != this);
```
This seems unnecessary, given that the docs for `GetReachableNodes` specifies "other". I think it should be clarified if `GetReachableNodes` should return the current node, since otherwise someone is going to forget this check somewhere.

ComicIronic commented on Content.Server/GameObjects/Components/NodeNetwork/INodeGroupFactory.cs:
```diff
@@ -0,0 +1,75 @@
+            var nodeGroupTypes = _reflectionManager.GetAllChildren<NodeGroup>();
+            foreach (var nodeGroupType in nodeGroupTypes)
+            {
+                var att = (NodeGroupAttribute) Attribute.GetCustomAttribute(nodeGroupType, typeof(NodeGroupAttribute));
```
You need `System.Reflection`, but this can be changed to `nodeGroupType.GetCustomAttribute<NodeGroupAttribute>()`, I believe.

collinlunn commented:
![0fa88ebfcbef4cb8a99a9480b1377a6b](https://user-images.githubusercontent.com/60152240/84954087-fe473180-b0b1-11ea-8462-dd02f6e7644c.png)

collinlunn commented:
Alright this basically all works. There are various bits that need improvements and code quality cleanup, but this PR has ballooned way to large, and I need to wrap this up.

ComicIronic commented on Content.Server/GameObjects/Components/Power/ApcNetComponents/PowerReceiverComponent.cs:
```diff
@@ -0,0 +1,211 @@
+                    if (provider.Connectable)
+                    {
+                        var distanceToProvider = provider.Owner.Transform.GridPosition.Distance(mapManager, Owner.Transform.GridPosition);
+                        if (distanceToProvider < Math.Min(PowerReceptionRange, provider.PowerTransferRange))
```
Is there a point to the `Min` here, since the entity is guaranteed to be in `PowerReceptionRange` by the query?

collinlunn replied:
The GetEntitiesInRange() query returns entities in a square range, so entities in the corner will be outside a circle of r = PowerReceptionRange.

ComicIronic commented on Content.Server/GameObjects/Components/Power/BatteryComponent.cs:
```diff
@@ -0,0 +1,95 @@
+        /// </summary>
+        public bool TryUseCharge(float chargeToUse)
+        {
+            if (chargeToUse > CurrentCharge)
```
`>=`? I realise that with floats the difference isn't so meaningful.

ComicIronic commented on Content.Server/GameObjects/Components/Power/BatteryComponent.cs:
```diff
@@ -0,0 +1,95 @@
+
+        private void SetCurrentCharge(float newChargeAmount)
+        {
+            var oldCharge = _currentCharge;
```
Is this unused?

ComicIronic commented on Content.Server/GameObjects/Components/Power/PowerNetComponents/PowerConsumerComponent.cs:
```diff
@@ -0,0 +1,78 @@
+        {
+            base.ExposeData(serializer);
+            serializer.DataField(ref _drawRate, "drawRate", 0);
+            serializer.DataField(ref _priority, "priorty", Priority.First);
```
Typo.

ComicIronic commented on Content.Server/GameObjects/Components/Power/WirePlacerComponent.cs:
```diff
@@ -23,6 +23,14 @@ internal class WirePlacerComponent : Component, IAfterInteract
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+            serializer.DataField(ref _wirePrototypeID, "wirePrototypeID", "HVWire");
```
I don't know that prototype IDs are the sanest way to go about this. PJB may have more to say.

collinlunn replied:
I'm not sure how else to handle it - any suggestions?

ComicIronic replied:
I'm not sure, personally. A proper enum for the different wire types seems like a better fit - since it seems like that's what this is actually being used for. That way, we can have different wire prototypes that still have the same wire type enum value.

PJB3005 replied:
I don't see what's wrong with just having this be a prototype ID, honestly.

ComicIronic replied:
My main concern was that this gets used in line 51 for checking prototype equality, which to me seems like almost an abuse of prototype IDs - since the actual code is trying to avoid creating multiple wires of the same type on the same tile.

collinlunn replied:
I understand your concern now, and have recoded the section that checks for a wire on a tile. Instead of comparing prototype ID, it looks for Nodes of a yaml-specified NodeGroupID set.

collinlunn replied:
I have implemented a WireType enum which is checked instead of the prototype ID.

ComicIronic commented:
I wonder if floats make sense - maybe we could get away with decimals or ints, given that power doesn't require arbitrary fractions.

PJB3005 commented:
Longs might be a good idea.

collinlunn commented:
I've removed most floats from the power code, and replaced them with ints. However, I've left BatteryComponent's CurrentCharge as a float. Since energy change in a battery per frame is calculated with an int energy consumption rate multiplied by a float frameTime from EntitySystem, leaving Charge as a float seemed to be the easiest way.

PJB3005 commented:
Bugs I ran into while testing this (more to follow, probably):

* Flashlight item status looks empty initially and only updates when you turn the flashlight on.

collinlunn commented:
Located and fixed all three of those issues.

ComicIronic commented on Content.Server/GameObjects/Components/NodeContainer/Nodes/NodeFactory.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace Content.Server.GameObjects.Components.NodeContainer.Nodes
+{
+    public class NodeFactory
```
Since you route this through IoC, I think it's a good idea to

  1. Make a `INodeFactory` interface, for mocking
  2. Slap doc comments on that interface

PJB3005 commented on Content.Server/GameObjects/Components/Power/PowerNetComponents/SmesComponent.cs:
```diff
@@ -0,0 +1,83 @@
+
+        private ChargeState _lastChargeState;
+
+        private DateTime _lastChargeStateChange;
```
Again, no using `DateTime`.

PJB3005 commented:
Architecture looks very clean and extensible in general, just a few minor nits and this seems good to merge.

collinlunn commented on Content.Server/GameObjects/Components/Power/ApcNetComponents/PowerReceiverComponent.cs:
```diff
@@ -0,0 +1,211 @@
+        ///     The fraction of APC charge below which this shuts off and stops using power.
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        public float PowerShutoffFraction { get => _powerShutoffFraction; set => SetPowerShutoffFraction(value); }
```
Removed PowerShutoffFraction

ComicIronic commented:
LGTM. Thanks for addressing all the reviews, I realise it can be annoying to have to fix a PR every so often as feedback trickles in.
